### PR TITLE
fix(api): permit lastAccessTime 0 in devices response

### LINKED ISSUE #1197: GET /account/devices will fail if lastAccessTime is zero
Response validation of `lastAccessTime` in the `/account/devices` endpoint requires that it is a positive number. Over in [the database migration](https://github.com/mozilla/fxa-auth-db-mysql/blob/master/lib/db/schema/patch-014-015.sql#L9), the column was created non-nullable but with a default value of `0`.

Pretty sloppy on my part, as I implemented both of those. :disappointed:

Anyway, I guess this should account for [jrgm's recent observation](https://github.com/mozilla/fxa-auth-server/pull/1131#issuecomment-188446941).

### PR DESCRIPTION
**Pull request by philbooth**
Fixes #1197.

The fix itself just updates the Joi data but I'm feeling pretty sketchy about the changes made to test it. As per https://github.com/mozilla/fxa-auth-server/issues/1197#issuecomment-188755474, it's difficult to force a certain value for `lastAccessTime` from the remote tests because it gets set automatically. I took the approach of passing in a query param to force it but there's one place in particular where that feels a little bit hairy, see my comment inline.

@rfk, @jrgm r?

### CODE REVIEW

File: lib/tokens/token.js [hunk 1]
```diff
@@ -60,7 +60,6 @@ module.exports = function (log, crypto, P, hkdf, Bundle, error) {
           Token.deriveTokenKeys(TokenType, bytes)
             .then(
               function (keys) {
-                details.createdAt = Date.now()
```
Comment (philbooth) [inline lib/tokens/token.js]: Here's the hairy bit. This line was forcing `createdAt` to always be `Date.now()`, rendering the assignment from `details.createdAt` on line 41 redundant. Does it bother anyone that I've removed it?

File: test/remote/device_tests.js [hunk 2]
```diff
@@ -230,7 +230,6 @@ TestServer.start(config)
             return client.updateDevice({ type: 'mobile' })
               .then(
                 function (r) {
-                  console.error('!!!!! PHIL !!!!! ' + Object.keys(r))
```
Comment (philbooth) [inline test/remote/device_tests.js]: No idea what this was doing here before, sorry. :blush:

File: lib/routes/account.js [hunk 3]
```diff
@@ -204,7 +204,8 @@ module.exports = function (
             email: account.email,
             emailCode: account.emailCode,
             emailVerified: account.emailVerified,
-            verifierSetAt: account.verifierSetAt
+            verifierSetAt: account.verifierSetAt,
+            createdAt: parseInt(request.query.createdAt)
```
Comment (rfk) [inline lib/routes/account.js]: We should be explicit about that happens if `request.query.createdAt` is not specified; IIUC this would currently produce `NaN` and we depend on the token-creation code to correctly interpret that as meaning "use the current time".

File: lib/tokens/token.js (hunk 1 above)
Comment (rfk) [inline lib/tokens/token.js]: It's not obvious to me why we were forcing that; with it removed, it just falls back to calling Date.now() in the `TokenType` constructor which calls Date.now() by default, right?

File: lib/routes/account.js (hunk 3 above)
Comment (philbooth) [inline lib/routes/account.js]: Good point, would it be clearer if I moved the call to `parseInt` later on, where the `>= 0` check is made?

Comment (rfk) [inline lib/routes/account.js]: I don't love exposing hooks like this in the public API, but we can probably get away with it in this case.  Some things we could consider to tighten it up:

* Don't respect it when `NODE_ENV` says we're in production mode
* Give it a more private-sounding name like `_createdAt`
* Only allow setting it to dates in the past, not the future

@philbooth thoughts?

Comment (rfk) [inline lib/routes/account.js]: > would it be clearer if I moved the call to parseInt later on

Or just something like `query.createdAt ? parseInt(query.createdAt) : undefined`

Comment (philbooth) [inline lib/routes/account.js]: > Don't respect it when `NODE_ENV` says we're in production mode

:+1:

> Give it a more private-sounding name like `_createdAt`

:+1:

> Only allow setting it to dates in the past, not the future

:+1:

> `query.createdAt ? parseInt(query.createdAt) : undefined`

:+1:

File: lib/tokens/token.js (hunk 1 above)
Comment (philbooth) [inline lib/tokens/token.js]: > with it removed, it just falls back to calling Date.now() in the `TokenType` constructor

Sorry, missed this last night. Yep, it still defaults to `Date.now()`.

**Comment by philbooth** [comment]
Thanks for the feedback @rfk, fixes pushed.

Parsing of the `_createdAt` param has now been extracted to a dedicated function, `optionallyOverrideCreatedAt`, and there are a couple more tests to cover the less-than-zero and in-the-future cases.

**Comment by philbooth** [comment]
Travis is saying one of the (unrelated) tests failed but, when I give it a kick to restart that build, nothing happens. Will try kicking it again later on.

File: lib/routes/account.js [hunk 4]
```diff
@@ -96,7 +96,8 @@ module.exports = function (
         var authPW = Buffer(form.authPW, 'hex')
         var locale = request.app.acceptLanguage
         var userAgentString = request.headers['user-agent']
-        var service = form.service || request.query.service
+        var query = request.query
```
Comment (rfk) [inline lib/routes/account.js]: tiny nit: for organizational purposes, I think this should live right next to the line that does `form = request.payload`.

File: lib/routes/account.js [hunk 5]
```diff
@@ -213,10 +215,17 @@ module.exports = function (
             )
         }
 
+        function optionallyOverrideCreatedAt () {
+          var createdAt = parseInt(query._createdAt)
+          if (createdAt < Date.now() && process.env.NODE_ENV !== 'production') {
```
Comment (rfk) [inline lib/routes/account.js]: There's a `config.isProduction` flag that can be used for this purpose (and whose definition leads me to believe that the appropriate `NODE_ENV` string is `"prod"`)

**Comment by rfk** [comment]
r+ with nits, thanks @philbooth
